go/internal/cbor/deterministic.go

```diff
@@ -0,0 +1,121 @@
+package cbor
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// Deterministic function loops through the given byte array containing CBOR sequence
+// and checks that they follow the deterministic principles described here:
+// https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c.
+func Deterministic(input []byte) error {
+	index := 0
+
+	for index < len(input) {
+		// TODO(sonkkeli): Remove debug print.
+		fmt.Printf("index=%v\n", index)
+
+		length, err := deterministicRec(input[index:])
+		if err != nil {
+			return err
+		}
+		index += length
+	}
+	return nil
+}
+
+// deterministicRec is a recursively called helper function to check the deterministicy of a bytes array.
+// TODO(sonkkeli): Recursively called parts will be the maps and arrays, which contain other types, meaning
+// unsigned integers, byte strings, text strings.
+func deterministicRec(input []byte) (int, error) {
+	cborItemLengthInBytes := 0
```

Comment:
I think it's easier to read if each `switch`-branch returns a value directly, instead of using this temporary variable. (You can still inject a debug print in the caller side.)